crates/core/c8y_api/src/http_proxy.rs

```diff
@@ -131,53 +135,120 @@ impl C8yEndPoint {
+    },
+
+    #[error("Error while parsing credentials file: '{0}': {1}.")]
+    TomlError(PathBuf, #[source] toml::de::Error),
 }
 
-impl C8yMqttJwtTokenRetriever {
+impl C8yAuthRetriever {
     pub fn from_tedge_config(
         tedge_config: &TEdgeConfig,
         c8y_profile: Option<&str>,
-    ) -> Result<Self, JwtRetrieverError> {
-        let mqtt_config = tedge_config
-            .mqtt_config()
-            .map_err(MqttConfigBuildError::from)?;
-
-        Ok(Self::new(
-            mqtt_config,
-            tedge_config
-                .c8y
-                .try_get(c8y_profile)?
-                .bridge
-                .topic_prefix
-                .clone(),
-        ))
+    ) -> Result<Self, C8yAuthRetrieverError> {
+        let c8y_config = tedge_config.c8y.try_get(c8y_profile)?;
+        let topic_prefix = c8y_config.bridge.topic_prefix.clone();
+
+        let auth = if c8y_config.use_basic_auth {
+            C8yAuthType::Basic {
+                credentials_path: c8y_config.credentials_path.clone(),
+            }
+        } else {
+            let mqtt_config = tedge_config
+                .mqtt_config()
+                .map_err(MqttConfigBuildError::from)?;
+
+            let topic = TopicFilter::new_unchecked(&format!("{topic_prefix}/s/dat"));
+            let mqtt_config = mqtt_config
+                .with_no_session() // Ignore any already published tokens, possibly stale.
+                .with_subscriptions(topic);
+
+            C8yAuthType::JwtToken {
+                mqtt_config: Box::new(mqtt_config),
+            }
+        };
+
+        Ok(Self { auth, topic_prefix })
     }
 
-    pub fn new(mqtt_config: mqtt_channel::Config, topic_prefix: TopicPrefix) -> Self {
-        let topic = TopicFilter::new_unchecked(&format!("{topic_prefix}/s/dat"));
-        let mqtt_config = mqtt_config
-            .with_no_session() // Ignore any already published tokens, possibly stale.
-            .with_subscriptions(topic);
+    pub async fn get_auth_header_value(&mut self) -> Result<HeaderValue, C8yAuthRetrieverError> {
+        let header_value = match &self.auth {
+            C8yAuthType::JwtToken { .. } => {
+                let jwt_token = self.get_jwt_token().await?;
+                format!("Bearer {}", jwt_token.token()).parse()?
+            }
+            C8yAuthType::Basic { .. } => {
+                let BasicCredentials { username, password } = self.get_basic_auth().await?;
+                format!("Basic {}", base64::encode(format!("{username}:{password}"))).parse()?
+            }
+        };
+        Ok(header_value)
+    }
 
-        C8yMqttJwtTokenRetriever {
-            mqtt_config,
-            topic_prefix,
-        }
+    async fn get_basic_auth(&self) -> Result<BasicCredentials, C8yAuthRetrieverError> {
+        let C8yAuthType::Basic { credentials_path } = &self.auth else {
+            panic!("This method should not be called if not intended to use Basic auth (username/password).");
+        };
```

Comment:
The fact that you have to add this `panic!` is a sign that something is wrong in the design of the `C8yAuthRetriever`.

Instead of `struct` using an `enum` field used to distinguish the cases, `C8yAuthRetriever` should directly be an `enum`:

```rust
pub enum C8yAuthRetriever {
    Basic { credentials_path: Utf8PathBuf },
    Jwt { mqtt_config: mqtt_channel::Config,  topic_prefix: TopicPrefix },
}

impl C8yAuthRetriever {
    pub async fn get_auth_header_value(&self) -> Result<HeaderValue, C8yAuthRetrieverError> {
        match self {
            C8yAuthRetriever::Jwt { mqtt_config, topic_prefix } => {
                let jwt_token = Self::get_jwt_token(mqtt_config, topic_prefix).await?;
                Ok(format!("Bearer {}", jwt_token))
            }
            C8yAuthRetriever::Basic { credentials_path } => {
                let basic_auth = Self::get_basic_auth(credentials_path).await?;
                Ok(format!("Basic {}", basic_auth))
            }
        };
    }
    
    async fn get_jwt_token(mqtt_config: mqtt_channel::Config,  topic_prefix: TopicPrefix) -> Result<String, C8yAuthRetrieverError> {
        ...
    }
    
    async fn get_basic_auth(credentials_path: Utf8PathBuf) -> Result<String, C8yAuthRetrieverError> {
        ...
    }
}
```

Note that, `get_basic_auth` is no more a method but a function which is given the expected parameters.

Reply:
Done in [0eeb299](https://github.com/thin-edge/thin-edge.io/pull/3196/commits/0eeb299ce6a4c7715f78839d222a82de44da0e78)